src/state.py

```diff
@@ -0,0 +1,119 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""The module for managing charm state."""
+
+import logging
+import os
+import typing
+from dataclasses import dataclass
+
+import ops
+from pydantic import BaseModel, Field, ValidationError
+
+import metadata
+
+# relation name used for compatibility with machine Jenkins server charm.
+SLAVE_RELATION = "slave"
+# agent relation name
+AGENT_RELATION = "agent"
+
+logger = logging.getLogger()
+
+
+class CharmStateBaseError(Exception):
+    """Represents error with charm state."""
+
+
+class InvalidStateError(CharmStateBaseError):
+    """Exception raised when state configuration is invalid."""
+
+    def __init__(self, msg: str = ""):
+        """Initialize a new instance of the InvalidStateError exception.
+
+        Args:
+            msg: Explanation of the error.
+        """
+        self.msg = msg
+
+
+class JenkinsConfig(BaseModel):
+    """The Jenkins config from juju config values.
+
+    Attrs:
+        server_url: The Jenkins server url.
+        agent_names: Custom Jenkins agent names.
+        agent_tokens: Tokens used to register Jenkins agent to Jenkins server.
+    """
+
+    server_url: str = Field(..., min_length=1)
+    agent_names: typing.List[str] = Field(..., min_items=1)
+    agent_tokens: typing.List[str] = Field(..., min_items=1)
+
+    @classmethod
+    def from_charm_config(cls, config: ops.ConfigData) -> typing.Optional["JenkinsConfig"]:
+        """Instantiate JenkinsConfig from charm config.
+
+        Args:
+            config: Charm configuration data.
+
+        Returns:
+            JenkinsConfig if configuration exists, None otherwise.
+        """
+        server_url = config.get("jenkins_url", None)
+        agent_names = config.get("jenkins_agent_name", None)
+        agent_tokens = config.get("jenkins_agent_token", None)
+        if not server_url and not agent_names and not agent_tokens:
+            return None
+        # type cast since Pydantic will throw validation error anyways.
+        return cls(
+            server_url=typing.cast(str, server_url),
+            agent_names=typing.cast(str, agent_names).split(":"),
```

Comment:
Are the pydantic checks run before or after this? If they are after the `.split` could fail if these are not `str`

Reply:
The dict mapping has to be str:str, I don't think it can be any other values :)

Reply:
Perhaps clear up the comment above? I read it as that pebble will raise an error, it sounds like there is a guarantee that it is a str?

Reply:
I meant pydantic not pebble

Reply:
Actually, you were right from the beginning, I've mistaken the dict mapping. Fixed, thank you!